nifi-nar-bundles/nifi-framework-bundle/nifi-stateless-bundle/nifi-stateless-engine/src/main/java/org/apache/nifi/stateless/repository/VolatileProvenanceRepository.java

```diff
@@ -0,0 +1,409 @@
+                           final IdentifierLookup idLookup) throws IOException {
+        if (initialized.getAndSet(true)) {
+            return;
+        }
+    }
+
+    @Override
+    public ProvenanceEventRepository getProvenanceEventRepository() {
+        return this;
+    }
+
+    @Override
+    public ProvenanceEventBuilder eventBuilder() {
+        return new StandardProvenanceEventRecord.Builder();
+    }
+
+    @Override
+    public void registerEvent(final ProvenanceEventRecord event) {
+        final long id = idGenerator.getAndIncrement();
+        ringBuffer.add(new IdEnrichedProvEvent(event, id));
+    }
+
+    @Override
+    public void registerEvents(final Iterable<ProvenanceEventRecord> events) {
+        for (final ProvenanceEventRecord event : events) {
+            registerEvent(event);
+        }
+    }
+
+    @Override
+    public List<ProvenanceEventRecord> getEvents(final long firstRecordId, final int maxRecords) throws IOException {
+        return getEvents(firstRecordId, maxRecords, null);
+    }
+
+    @Override
+    public List<ProvenanceEventRecord> getEvents(final long firstRecordId, final int maxRecords, final NiFiUser user) throws IOException {
+        return ringBuffer.getSelectedElements(new RingBuffer.Filter<ProvenanceEventRecord>() {
+            @Override
+            public boolean select(final ProvenanceEventRecord value) {
+                if (!isAuthorized(value, user)) {
+                    return false;
+                }
+
+                return value.getEventId() >= firstRecordId;
+            }
+        }, maxRecords);
+    }
+
+    @Override
+    public Long getMaxEventId() {
+        final ProvenanceEventRecord newest = ringBuffer.getNewestElement();
+        return (newest == null) ? null : newest.getEventId();
+    }
+
+    public ProvenanceEventRecord getEvent(final String identifier) throws IOException {
+        final List<ProvenanceEventRecord> records = ringBuffer.getSelectedElements(new RingBuffer.Filter<ProvenanceEventRecord>() {
+            @Override
+            public boolean select(final ProvenanceEventRecord event) {
+                return identifier.equals(event.getFlowFileUuid());
+            }
+        }, 1);
+        return records.isEmpty() ? null : records.get(0);
+    }
+
+    @Override
+    public ProvenanceEventRecord getEvent(final long id) {
+        final List<ProvenanceEventRecord> records = ringBuffer.getSelectedElements(new RingBuffer.Filter<ProvenanceEventRecord>() {
+            @Override
+            public boolean select(final ProvenanceEventRecord event) {
+                return event.getEventId() == id;
+            }
+        }, 1);
+
+        return records.isEmpty() ? null : records.get(0);
+    }
+
+    @Override
+    public ProvenanceEventRecord getEvent(final long id, final NiFiUser user) {
+        final ProvenanceEventRecord event = getEvent(id);
+        if (event == null) {
+            return null;
+        }
+
+        authorize(event, user);
```

Comment:
I think we can avoid even calling authorize, no?

Reply:
Agree